lib/active_merchant/billing/gateways/tsys_multipass.rb

```diff
@@ -0,0 +1,190 @@
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(/\"deviceID\":\K(?:(?!,).)*/, '[FILTERED]').
+          gsub(/\"transactionKey\":\K(?:(?!,).)*/, '[FILTERED]').
+          gsub(/\"cardNumber\":\K(?:(?!,).)*/, '[FILTERED]').
+          gsub(/\"expirationDate\":\K(?:(?!,).)*/, '[FILTERED]').
+          gsub(/\"token\":\K(?:(?!,).)*/, '[FILTERED]')
+      end
+
+      private
+
+      def request_params(options)
+        {
+          "deviceID": @options[:device_id],
+          "transactionKey": @options[:transaction_key]
+        }.merge!(options)
+      end
+
+      def commit(request_body:)
+        @response =
+          Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |https|
+            request      = Net::HTTP::Post.new(uri, {'Content-Type' => CONTENT_TYPE })
+            request.body = request_body
+            # Making the call
+            https.request(request)
+          end
+
+        # Parsing the response body
+        @parsed_body = parse(response.body)
+
+        Response.new(
+          success?,
+          message,
+          parsed_body,
+          amount: amount,
+          error_code: error_code,
+          authorization: authorization,
+          test: test?
+        )
+      end
+
+      def success?
+        return false unless recognized_response_root_key?
+
+        parsed_body_root_value['status'] == 'PASS'
+      end
+
+      def message
+        return BLANK unless recognized_response_root_key?
+
+        parsed_body_root_value['responseMessage']
+      end
+
+      def error_code
+        return BLANK unless error?
+
+        response_code
+      end
+
+      def error?
+        !response_code.start_with?('A')
+      end
+
+      def response_code
+        return BLANK unless recognized_response_root_key?
+
+        parsed_body_root_value['responseCode']
+      end
+
+      def authorization
+        return BLANK unless recognized_response_root_key?
+
+        parsed_body_root_value['transactionID']
+      end
+
+      def amount
+        return BLANK unless recognized_response_root_key?
+
+        parsed_body_root_value[amount_key_mapping[parsed_body_root_key]]
+      end
+
+      # This method gives us mapping of which amount field to
+      # fetch based on the transaction response types.
+      def amount_key_mapping
+        {
+          'SaleResponse' => 'processedAmount',
+          'AuthResponse' => 'processedAmount',
+          'CaptureResponse' => 'transactionAmount',
+          'VoidResponse' => 'voidedAmount',
+          'ReturnResponse' => 'returnedAmount'
+        }
+      end
+
+      def recognized_response_root_key?
+        @recognized_response_root_key ||=
+          WHITELISTED_RESPONSE_ROOT_KEYS.include?(parsed_body_root_key)
+      end
+
+      def parsed_body_root_key
+        parsed_body.first&.first
+      end
+
+      def parsed_body_root_value
+        (parsed_body.present? && parsed_body.first[1]) || EMPTY_OBJ
+      end
+
+      def parse(body)
```

Comment:
I took a look at the `active_merchant` and I noticed that other gateways are parsing JSON just like you're doing. I'd probably recommend isolating this and make a global helper for the gem.

Some examples:
https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/adyen.rb#L443
https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/balanced.rb

<img width="1440" alt="Screen Shot 2019-12-16 at 11 38 50 AM" src="https://user-images.githubusercontent.com/1292556/70937160-b1967700-1ff8-11ea-8152-d5c043f9ef22.png">

It isn't a mandatory change. It's a nice to have but I believe that'd be a plus for this gem if we have a `Parser` object that is able to parse `JSON`, `XML` and `CGI`. These are the ones that I saw. I'd start with `JSON.`

Reply:
Ill leave this for now.